src/ReservoirPriceOracle.sol

```diff
@@ -1,32 +1,405 @@
+    function getLatest(OracleLatestQuery calldata aQuery) external view /*nonReentrant*/ returns (uint256) {
+        (address lToken0, address lToken1) = aQuery.base.sortTokens(aQuery.quote);
+        ReservoirPair lPair = pairs[lToken0][lToken1];
+        _validatePair(lPair);
+
+        (,,, uint256 lIndex) = lPair.getReserves();
+        uint256 lResult = lPair.getInstantValue(aQuery.variable, lIndex, lToken0 == aQuery.quote);
+        return lResult;
     }
 
+    /// @inheritdoc IReservoirPriceOracle
     function getPastAccumulators(OracleAccumulatorQuery[] memory aQueries)
         external
         view
-        returns (int256[] memory rResults)
-    { }
+        returns (
+            // nonReentrant
+            int256[] memory rResults
+        )
+    {
+        rResults = new int256[](aQueries.length);
+
+        OracleAccumulatorQuery memory query;
+        for (uint256 i = 0; i < aQueries.length; ++i) {
+            query = aQueries[i];
+            (address lToken0, address lToken1) = query.base.sortTokens(query.quote);
+            ReservoirPair lPair = pairs[lToken0][lToken1];
+            _validatePair(lPair);
+
+            (,,, uint16 lIndex) = lPair.getReserves();
+            int256 lAcc = lPair.getPastAccumulator(query.variable, lIndex, query.ago);
+            // safety: negation will not overflow as the accumulator's type is int88
+            rResults[i] = lToken0 == query.base ? lAcc : -lAcc;
+        }
+    }
+
+    /// @inheritdoc IReservoirPriceOracle
+    function getLargestSafeQueryWindow() external pure returns (uint256) {
+        return Buffer.SIZE;
+    }
+
+    ///////////////////////////////////////////////////////////////////////////////////////////////
+    //                                 INTERNAL FUNCTIONS                                        //
+    ///////////////////////////////////////////////////////////////////////////////////////////////
+
+    function _validatePair(ReservoirPair aPair) internal pure {
+        if (address(aPair) == address(0)) revert NoDesignatedPair();
+    }
+
+    // TODO: replace this with safe, audited lib function
+    function _calcPercentageDiff(uint256 aOriginal, uint256 aNew) internal pure returns (uint256) {
+        if (aOriginal == 0) return 0;
+
+        if (aOriginal > aNew) {
+            return (aOriginal - aNew) * 1e18 / aOriginal;
+        } else {
+            return (aNew - aOriginal) * 1e18 / aOriginal;
+        }
+    }
+
+    function _rewardUpdater(address lRecipient) internal {
+        // N.B. Revisit this whenever deployment on a new chain is needed
+        // we use `block.basefee` instead of `ArbGasInfo::getMinimumGasPrice()` on ARB because the latter will always return
+        // the demand insensitive base fee, while the former can return real higher fees during times of congestion
+        // safety: this mul will not overflow even in extreme cases of `block.basefee`
+        uint256 lPayoutAmt = block.basefee * rewardMultiplier;
```

Comment:
We should call rewardMultiplier rewardGasAmount

Reply:
Technically there may be a need to split the reward into:

- base reward
- per price reward

as updating a single price costs more than updating N where N > 1

Reply:
> We should call rewardMultiplier rewardGasAmount

okay makes sense. I've been looking for a good name. Changed variable name

Reply:
> Technically there may be a need to split the reward into:

okay I think we can have 2 variables: 

`baseRewardGasAmount` and `additionalRewardGasAmount`

If N == 1, only pay out baseRewardGasAmount, if N == 2 || 3:
`total payout = baseRewardGasAmount + additionalRewardGasAmount * (N-1)` 

Agree? 

This `additionalRewardGasAmount` will be of a uint64 type as well. Then the 4 `uint64` variables will fit into one storage slot

Reply:
Before committing to this I think we need to discuss on a call the economics here. It looks like `priceDeviationThreshold` is global and not per price. This is more gas efficient but I wonder if there's merit in different thresholds for different assets (like stablecoins vs volatile alts).

Reply:
In the call 2 weeks ago we decided that if the money market could re-specify the oracle after deployment we don't need to implement this right now. 

However, according to Euler, it seems that the address of the oracle specified by the vaults will be fixed and will not be changeable. 

So I think we have 2 options: 

1. Make our oracle contract an upgradeable proxy, updating as needed
2. Build all the logic we need now, so that we need not upgrade the oracle ever. 

I'm leaning towards option 1.

Reply:
After discussions, we have decided that we will not make it into an upgradeable proxy, and that we will introduce a fallback oracle like what `EulerRouter` has. 

If changes are needed, they will be implemented in MM V2

Reply:
At worst, we can point the `fallbackOracle` to an updated instance of our oracle, while clearing all routes for this oracle. This will forward queries to the `fallbackOracle`, at the cost of an additional call and an extra `SLOAD`